contracts/src/Teleporter/TeleporterMessenger.sol

```diff
@@ -142,17 +142,17 @@ contract TeleporterMessenger is ITeleporterMessenger, ReentrancyGuards {
         TeleporterMessage calldata message
     ) external senderNonReentrant {
         // Get the previously sent message hash.
-        bytes32 messageHash = sentMessageInfo[destinationChainID][
-            message.messageID
-        ].messageHash;
-        if (messageHash == bytes32(0)) {
+        SentMessageInfo memory existingMessageInfo = sentMessageInfo[
+            destinationChainID
+        ][message.messageID];
```

Comment:
Do you know if there is any significant gas difference in bringing the entire `SentMessageInfo` into `memory`? I'm not sure how well the Solidity compiler is able to optimize the existing case of accessing the `messageHash` inline.

Reply:
Good question. 

When running `forge test --gas-report` on the current version in `main`, I get:
```
| src/Teleporter/TeleporterMessenger.sol:TeleporterMessenger contract |                 |        |        |        |         |
|---------------------------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                                     | Deployment Size |        |        |        |         |
| 2174481                                                             | 10854           |        |        |        |         |
| Function Name                                                       | min             | avg    | median | max    | # calls |
| addFeeAmount                                                        | 3486            | 5545   | 4070   | 9507   | 7       |
| checkIsAllowedRelayer                                               | 950             | 1121   | 950    | 1643   | 5       |
| checkRelayerRewardAmount                                            | 764             | 764    | 764    | 764    | 8       |
| getFeeInfo                                                          | 851             | 1651   | 851    | 4851   | 5       |
| getMessageHash                                                      | 643             | 928    | 643    | 2643   | 7       |
| getNextMessageID                                                    | 660             | 1993   | 2660   | 2660   | 3       |
| getReceiptAtIndex                                                   | 903             | 1038   | 903    | 1616   | 21      |
| getReceiptQueueSize                                                 | 730             | 3896   | 4730   | 4730   | 24      |
| getRelayerRewardAddress                                             | 629             | 754    | 629    | 2629   | 16      |
| messageReceived                                                     | 673             | 839    | 673    | 2673   | 12      |
| receiveCrossChainMessage                                            | 609             | 87978  | 101991 | 178925 | 51      |
| redeemRelayerRewards                                                | 1946            | 2230   | 2008   | 2736   | 3       |
| retryMessageExecution                                               | 3748            | 16726  | 9318   | 57861  | 11      |
| retryReceipts                                                       | 8634            | 37784  | 40410  | 64310  | 3       |
| retrySendCrossChainMessage                                          | 5440            | 7799   | 7753   | 10206  | 3       |
| sendCrossChainMessage                                               | 34540           | 101038 | 111478 | 137275 | 32      |
```

And when running it on this branch, I get:
```
| src/Teleporter/TeleporterMessenger.sol:TeleporterMessenger contract |                 |       |        |        |         |
|---------------------------------------------------------------------|-----------------|-------|--------|--------|---------|
| Deployment Cost                                                     | Deployment Size |       |        |        |         |
| 2215325                                                             | 11058           |       |        |        |         |
| Function Name                                                       | min             | avg   | median | max    | # calls |
| addFeeAmount                                                        | 3486            | 5547  | 4070   | 9513   | 7       |
| checkIsAllowedRelayer                                               | 954             | 1125  | 954    | 1647   | 5       |
| checkRelayerRewardAmount                                            | 764             | 764   | 764    | 764    | 8       |
| getFeeInfo                                                          | 851             | 1651  | 851    | 4851   | 5       |
| getMessageHash                                                      | 643             | 928   | 643    | 2643   | 7       |
| getNextMessageID                                                    | 660             | 1993  | 2660   | 2660   | 3       |
| getReceiptAtIndex                                                   | 903             | 1038  | 903    | 1616   | 21      |
| getReceiptQueueSize                                                 | 730             | 3896  | 4730   | 4730   | 24      |
| getRelayerRewardAddress                                             | 629             | 754   | 629    | 2629   | 16      |
| messageReceived                                                     | 673             | 839   | 673    | 2673   | 12      |
| receiveCrossChainMessage                                            | 609             | 88430 | 102587 | 179522 | 51      |
| redeemRelayerRewards                                                | 1952            | 2234  | 2014   | 2736   | 3       |
| retryMessageExecution                                               | 3748            | 16664 | 9247   | 57790  | 11      |
| retryReceipts                                                       | 8634            | 38256 | 41118  | 65018  | 3       |
| retrySendCrossChainMessage                                          | 5692            | 9533  | 10824  | 12085  | 3       |
| sendCrossChainMessage                                               | 34546           | 99304 | 108115 | 137991 | 32      |
```

Not very significant either, but it seems the take aways are that:
- Declaring the `SentMessageInfo memory` variable here costs more gas than just a `bytes32` value. This makes sense to me. The resulting gas cost are still reasonable IMO.
- Declaring the `TeleporterFeeInfo` as `memory` in `_sendCrossChainMessage` rather than constructing it inline seems to be an incremental gas savings. This is surprising to me, but I'll take it 😅. Need to look up why it is the case.